### PR TITLE
Account for memory usage in doc values aggregators.

### PR DESCRIPTION
**Pull request by kovrus**
## Summary of the changes / Why this improves CrateDB

See commits.

It should supersede https://github.com/crate/crate/pull/10396

## Checklist

 - [ ] Added an entry in `CHANGES.txt` for user facing changes
 - [ ] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Comment by kovrus** [comment]
@seut @mfussenegger could you please take a look at this approach? So the RamAccounting is used only when the state is initialized, for the collect set and array aggregation there would be anyways the full concrete implementation of the `DocValueAggregator` where memory also will be accounted in `DocValueAggregator#apply`.

**Review by mfussenegger — COMMENTED** [review]
Looks good to me

**Comment by seut** [comment]
lgtm

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/ArbitraryAggregation.java [hunk 1]
```diff
@@ -198,21 +203,26 @@ public Object partialResult(RamAccounting ramAccounting, MutableObject state) {
...
+            this.sizeEstimator = sizeEstimator;
         }
 
         @Nullable
         @Override
         public Object partialResult(RamAccounting ramAccounting, MutableObject state) {
             if (state.hasValue()) {
-                return columnDataType.sanitizeValue(state.value());
+                var partialResult = columnDataType.sanitizeValue(state.value());
+                ramAccounting.addBytes(sizeEstimator.estimateSize(partialResult));
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/execution/engine/aggregation/impl/ArbitraryAggregation.java:224]: Shouldn't we do this in `apply` instead, to account the used bytes earlier? Or does that make the abstractions more complex?

Comment (kovrus) [reply server/src/main/java/io/crate/execution/engine/aggregation/impl/ArbitraryAggregation.java:224]: Yes, forgot that it is called only once. I will just override the `apply` then and account for memory there.

Comment (kovrus) [reply server/src/main/java/io/crate/execution/engine/aggregation/impl/ArbitraryAggregation.java:224]: Hm it makes it a bit hard, it is not possible to reuse the abstraction, and extending it does not make sense. We need to have 4 different separate implementations for this case :/

Comment (mfussenegger) [reply server/src/main/java/io/crate/execution/engine/aggregation/impl/ArbitraryAggregation.java:224]: I'm a bit worried for the GROUP BY cases if we delay the accounting to `partialResult`. For aggregations it doesn't matter as it is only 1 row anyways. But afaik in the GROUP BY case the `partial` calls would happen at the very end, so we could have `unique_key_values * num_arbitrary_aggregation` aggregation states un-accounted for in-memory during the initial grouping phase, before any `arbitrary` state is accounted for.

Comment (kovrus) [reply server/src/main/java/io/crate/execution/engine/aggregation/impl/ArbitraryAggregation.java:224]: I see, let me look a bit more at it, but think for the arbitrary that's only the way to go with separate implementations or have smth like `ThreeConsumer` :)

**Comment by kovrus** [comment]
not related failure `org.elasticsearch.index.engine.InternalEngineTests.testTrackMaxSeqNoOfNonAppendOnlyOperations`

**Review by mfussenegger — APPROVED** [review]

**Pull request merged**